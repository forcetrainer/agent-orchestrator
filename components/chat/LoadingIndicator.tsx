'use client';

import { memo } from 'react';

/**
 * LoadingIndicator Component
 *
 * Displays animated loading indicator when agent is processing
 * Story 3.6 - Task 1: Create LoadingIndicator component with animation
 *
 * AC-6.2: Indicator shows "Agent is thinking..." or similar message
 * AC-6.5: Visual cue is clear but not distracting (subtle animation)
 *
 * Styling matches agent message appearance (left-aligned, gray background)
 * Animation: CSS-based animated dots (1.4s loop) for performance
 * Accessibility: aria-live="polite" for screen reader support
 */
export const LoadingIndicator = memo(function LoadingIndicator() {
  return (
    <div
      className="max-w-[75%] mr-auto bg-gray-200 text-gray-900 px-4 py-3 rounded-lg"
      role="status"
      aria-live="polite"
      aria-label="Agent is thinking"
    >
      <div className="flex items-center gap-2">
        <span className="text-gray-700">Agent is thinking</span>
        <div className="flex gap-1">
          <span className="inline-block animate-bounce" style={{ animationDelay: '0s' }}>.</span>
          <span className="inline-block animate-bounce" style={{ animationDelay: '0.1s' }}>.</span>
          <span className="inline-block animate-bounce" style={{ animationDelay: '0.2s' }}>.</span>
        </div>
      </div>
    </div>
  );
});
