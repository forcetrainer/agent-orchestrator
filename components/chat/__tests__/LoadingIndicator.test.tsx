import { render, screen } from '@testing-library/react';
import { LoadingIndicator } from '../LoadingIndicator';

/**
 * LoadingIndicator Component Tests
 * Story 3.6 - Task 5: Unit tests for LoadingIndicator component
 *
 * AC-6.2: Indicator shows "Agent is thinking..." or similar message
 * AC-6.5: Visual cue is clear but not distracting (subtle animation)
 */
describe('LoadingIndicator', () => {
  // Subtask 5.1: Test component renders with "Agent is thinking..." text
  it('renders with "Agent is thinking" text', () => {
    render(<LoadingIndicator />);
    expect(screen.getByText(/Agent is thinking/i)).toBeInTheDocument();
  });

  // Subtask 5.2: Test animation is present and loops continuously
  it('has animated dots with bounce animation', () => {
    const { container } = render(<LoadingIndicator />);
    const dots = container.querySelectorAll('.animate-bounce-dot');
    expect(dots).toHaveLength(3);
  });

  // Subtask 5.2: Test animation delays are staggered
  it('has staggered animation delays on dots', () => {
    const { container } = render(<LoadingIndicator />);
    const dots = container.querySelectorAll('.animate-bounce-dot');

    expect(dots[0]).toHaveClass('animation-delay-0');
    expect(dots[1]).toHaveClass('animation-delay-200');
    expect(dots[2]).toHaveClass('animation-delay-400');
  });

  // Subtask 5.3: Test ARIA live region for accessibility
  it('has aria-live="polite" for screen reader support', () => {
    const { container } = render(<LoadingIndicator />);
    const loadingDiv = container.querySelector('[aria-live="polite"]');
    expect(loadingDiv).toBeInTheDocument();
  });

  // Subtask 5.3: Test role and aria-label for accessibility
  it('has role="status" and aria-label for accessibility', () => {
    const { container } = render(<LoadingIndicator />);
    const loadingDiv = container.querySelector('[role="status"]');
    expect(loadingDiv).toBeInTheDocument();
    expect(loadingDiv).toHaveAttribute('aria-label', 'Agent is thinking');
  });

  // Subtask 5.4: Test styling matches agent message appearance
  it('matches agent message styling (left-aligned, gray background)', () => {
    const { container } = render(<LoadingIndicator />);
    const loadingDiv = container.querySelector('.bg-gray-200');

    // Check for agent message styles
    expect(loadingDiv).toBeInTheDocument();
    expect(loadingDiv).toHaveClass('mr-auto'); // Left-aligned
    expect(loadingDiv).toHaveClass('max-w-[75%]'); // Max width like messages
    expect(loadingDiv).toHaveClass('rounded-lg'); // Rounded corners
    expect(loadingDiv).toHaveClass('px-4', 'py-3'); // Padding like messages
  });

  // Subtask 5.5: Snapshot test for consistent rendering
  it('matches snapshot', () => {
    const { container } = render(<LoadingIndicator />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
