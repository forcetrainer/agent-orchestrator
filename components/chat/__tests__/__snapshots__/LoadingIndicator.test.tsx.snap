// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`LoadingIndicator matches snapshot 1`] = `
<div
  aria-label="Agent is thinking"
  aria-live="polite"
  class="max-w-[75%] mr-auto bg-gray-200 text-gray-900 px-4 py-3 rounded-lg"
  role="status"
>
  <div
    class="flex items-center gap-2"
  >
    <span
      class="text-gray-700"
    >
      Agent is thinking
    </span>
    <div
      class="flex gap-1"
    >
      <span
        class="animate-bounce-dot animation-delay-0"
      >
        .
      </span>
      <span
        class="animate-bounce-dot animation-delay-200"
      >
        .
      </span>
      <span
        class="animate-bounce-dot animation-delay-400"
      >
        .
      </span>
    </div>
  </div>
</div>
`;
